lib/srv/sessiontracker.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package srv
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	apidefaults "github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/trace"
+)
+
+// SessionTracker is a session tracker for a specific session. It tracks
+// the session in memory and broadcasts updates to the given service (backend).
+type SessionTracker struct {
+	closeC chan struct{}
+	// tracker is the in memory session tracker
+	tracker types.SessionTracker
+	// trackerCond is used to provide synchronized access to tracker
+	// and to broadcast state changes.
+	trackerCond *sync.Cond
+	// service is used to share session tracker updates with the service
+	service services.SessionTrackerService
+}
+
+// NewSessionTracker returns a new SessionTracker for the given types.SessionTracker
+func NewSessionTracker(ctx context.Context, trackerSpec types.SessionTrackerSpecV1, service services.SessionTrackerService) (*SessionTracker, error) {
+	if service == nil {
+		return nil, trace.BadParameter("missing parameter service")
+	}
+
+	t, err := types.NewSessionTracker(trackerSpec)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if t, err = service.CreateSessionTracker(ctx, t); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &SessionTracker{
+		service:     service,
+		tracker:     t,
+		trackerCond: sync.NewCond(&sync.Mutex{}),
+		closeC:      make(chan struct{}),
+	}, nil
+}
+
+// Close closes the session tracker and sets the tracker state to terminated
+func (s *SessionTracker) Close(ctx context.Context) error {
+	close(s.closeC)
+	err := s.UpdateState(ctx, types.SessionState_SessionStateTerminated)
+	return trace.Wrap(err)
+}
+
+const sessionTrackerExpirationUpdateInterval = apidefaults.SessionTrackerTTL / 6
+
+// UpdateExpirationLoop extends the session tracker expiration by 1 hour every 10 minutes
+// until the SessionTracker or ctx is closed.
+func (s *SessionTracker) UpdateExpirationLoop(ctx context.Context, clock clockwork.Clock) error {
+	for {
+		select {
+		// We use clock.After rather than clock.Ticker here because ticker
+		// does not work with clock.BlockUntil, which is useful in tests.
+		case time := <-clock.After(sessionTrackerExpirationUpdateInterval):
```

Comment:
If `clock.After` is the same as `time.After` then it will be leaking channels leading to memory leaks. Can we use ticker here?

Reply:
Yeah you're right, that's unfortunate as it makes `clock.BlockUntil` useless, but I'll alter the tests to work with ticker.